sdk/core/src/request.rs

```diff
@@ -148,3 +166,140 @@ impl Request {
         self.insert_header(item.name(), item.value());
     }
 }
+
+/// The body content of a service client request.
+/// This allows callers to pass a model to serialize or raw content to client methods.
+#[derive(Clone, Debug)]
+pub struct RequestContent<T> {
+    body: Body,
+    phantom: PhantomData<T>,
+}
+
+impl<T> RequestContent<T> {
+    /// Gets the body of the request.
+    pub fn body(&self) -> &Body {
+        &self.body
+    }
+
+    /// Create a new `RequestContent` from byte slice.
+    pub fn from(bytes: Vec<u8>) -> Self {
+        Self {
+            body: Body::Bytes(Bytes::from(bytes)),
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[cfg(test)]
+impl<T> PartialEq for RequestContent<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.body.eq(&other.body)
+    }
+}
+
+impl<T> From<RequestContent<T>> for Body {
+    fn from(content: RequestContent<T>) -> Self {
+        content.body
+    }
+}
+
+impl<T> TryFrom<Bytes> for RequestContent<T> {
+    type Error = crate::Error;
+    fn try_from(body: Bytes) -> Result<Self, Self::Error> {
+        Ok(Self {
+            body: Body::Bytes(body),
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<T> TryFrom<Vec<u8>> for RequestContent<T> {
+    type Error = crate::Error;
+    fn try_from(body: Vec<u8>) -> Result<Self, Self::Error> {
+        Ok(Self {
+            body: Bytes::from(body).into(),
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<T> TryFrom<&'static str> for RequestContent<T> {
+    type Error = crate::Error;
+    fn try_from(body: &'static str) -> Result<Self, Self::Error> {
+        Ok(Self {
+            body: Bytes::from_static(body.as_bytes()).into(),
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<T> FromStr for RequestContent<T> {
+    type Err = crate::Error;
+    fn from_str(body: &str) -> Result<Self, Self::Err> {
+        let body: Bytes = Bytes::copy_from_slice(body.as_bytes());
+        Ok(Self {
+            body: Body::Bytes(body),
+            phantom: PhantomData,
+        })
+    }
+}
+
+#[cfg(test)]
```

Comment:
Food for thought: 
Tests in the file is common and an idiomatic pattern. I have also seen patterns for putting tests in *_test.rs files. It might be cleaner to have that model.

Reply:
FWIW, there are strong technical reasons for tests in the same source file - specifically, tests in the same source file can access private fields in structures.

Typically, I've seen the _test.rs files used for integration or non-unit tests, and the `#[cfg(test)]` in source files used for unit tests.

Reply:
Within the same folder you can get the same via use of `super`.

Reply:
Tests in separate files are generally integration tests: https://doc.rust-lang.org/book/ch11-03-test-organization.html

If you put them in separate files, they are separate modules and forces your non-test module to make everything public at least to the crate, which is not always what you want. Submodules have access to ancestors' privates.

I think we should remain consistent.

FWIW, you don't even need the ol'

```rust
#[cfg(test)]
mod tests {
  use super::*;
}
```

It's an easy way to condition things, but it's not uncommon to see tests interspersed with functions, since the `#[test]` procedural macro automatically conditions on `test`:

```rust
fn foo() -> bool {
  true
}

#[test]
fn test_foo() {
  assert!(foo());
}

fn bar() -> &'static str {
  "hello, world!"
}

#[test]
fn test_bar() {
  assert_eq!("hello, world!", bar());
}
```

In function-heavy files, this can actually be handy.